rfcs/0000-graphql-api.md

```diff
@@ -0,0 +1,431 @@
+  privateKey: PrivateKey!
+  balance(consensus: ConsensusStatus): UInt64!
+}
+
+type NodeStatus {
+  network: String
+}
+
+## Input types
+
+input AddWalletInput {
+  public: PublicKey
+  private: PrivateKey
+}
+
+input DeleteWalletInput {
+  public: PublicKey
+}
+
+input AddPaymentReceiptInput {
+  receipt: String
+}
+
+input SetNetworkInput {
+  address: String
+}
+
+input SetSnarkWorkerInput {
+  worker: PublicKey!
+  fee: UInt32!
+}
+
+input CreatePaymentInput {
+  from: PublicKey!,
+  to: PublicKey!,
+  amount: UInt64!,
+  fee: UInt32!,
+  memo: String
+}
+
+input PaymentFilterInput {
+  toOrFrom: PublicKey,
+}
+
+input BlockFilterInput {
+  creator: PublicKey,
+}
+
+## Payload types
+
+type CreatePaymentPayload {
+  payment: Payment
+}
+
+type SetSnarkWorkerPayload {
+  worker: SnarkWorker
+}
+
+type SetNetworkPayload {
+  address: String
+}
+
+type AddPaymentReceiptPayload {
+  payment: Payment
+}
+
+type AddWalletPayload {
+  publicKey: PublicKey
+}
+
+type DeleteWalletPayload {
+  publicKey: PublicKey
+}
+
+# Pagination types
+
+type PageInfo {
+  hasPreviousPage: Boolean!
+  hasNextPage: Boolean!
+}
+
+type PaymentEdge {
+  cursor: String
+  node: PaymentUpdate
+}
+
+type PaymentConnection {
+  edges: [PaymentEdge]
+  nodes: [PaymentUpdate]
+  pageInfo: PageInfo!
+  totalCount: Int
+}
+
+type BlockEdge {
+  cursor: String
+  node: BlockUpdate
+}
+
+type BlockConnection {
+  edges: [BlockEdge]
+  nodes: [BlockUpdate]
+  pageInfo: PageInfo!
+  totalCount: Int
+}
+
+type Query {
+  # List of wallets currently tracked by the node
+  wallets: [Wallet!]!
+  
+  # Gets balance of key at a certain consensus state
+  # Note: `consensus` is optional as we will likely decide one
+  # state to be the "real" balance
+  balance(publicKey: PublicKey!, consensus: ConsensusStatus): UInt64!
+  
+  payments(
+    filter: PaymentFilterInput,
+    first: Int,
+    after: String,
+    last: Int,
+    before: String): PaymentConnection
+  
+  blocks(
+    filter: BlockFilterInput,
+    first: Int,
+    after: String,
+    last: Int,
+    before: String): BlockConnection
+  
+  # Null if node isn't performing snark work
+  currentSnarkWorker: SnarkWorker
+
+  # Current sync status of the node
+  syncState: SyncUpdate!
+  
+  # version of the node (commit hash or version #)
+  version: String!
+  
+  # Network that the node is connected to
+  network: String
+  status: NodeStatus
+}
+
+type Mutation {
+  createPayment(input: CreatePaymentInput!): CreatePaymentPayload
+  
+  setSnarkWorker(input: SetSnarkWorkerInput!): SetSnarkWorkerPayload
+  
+  # Configure which network your node is connected to
+  setNetwork(input: SetNetworkInput!): SetNetworkPayload
+  
+  # Adds transaction to the node (note: Not sure how we want to represent this yet)
+  addPaymentReceipt(input: AddPaymentReceiptInput!): AddPaymentReceiptPayload
+  
+  # Tell server to track a private key and all associated transactions
+  addWallet(input: AddWalletInput!): AddWalletPayload
```

Comment:
`AddWalletInput` should probably just be handled server-side

Reply:
As discussed, public_key is nullable so we should be fine. For now, we'll only support passing null there on the backend though